buttercup/cogs/heatmap.py

```diff
@@ -164,6 +263,112 @@ async def _heatmap(
+        description="Display the yearly activity map for the given user.",
+        options=[
+            create_option(
+                name="username",
+                description="The user to get the activity map for.",
+                option_type=3,
+                required=False,
+            ),
+            create_option(
+                name="before",
+                description="The end time of the activity map, "
+                "it will show one year before this date.",
+                option_type=3,
+                required=False,
+            ),
+        ],
+    )
+    async def activity_map(
+        self,
+        ctx: SlashContext,
+        username: Optional[str] = "me",
+        before: Optional[str] = None,
+    ) -> None:
+        """Generate a yearly activity heatmap for the given user."""
+        start = datetime.now()
+
+        # First parse the end time for the activity map
+        _, before_time, _ = parse_time_constraints(None, before)
+
+        # Then calculate the start time (one year before the end)
+        after = (
+            "1 year"
+            if before_time is None
+            else (before_time - timedelta(days=365)).date().isoformat()
+        )
+        # and get the string representing the time span
+        after_time, _, time_str = parse_time_constraints(after, before)
+
+        msg = await ctx.send(
+            i18n["activity"]["getting_activity"].format(
+                user=get_initial_username(username, ctx)
+            )
+        )
+
+        from_str = after_time.isoformat() if after_time else None
+        until_str = before_time.isoformat() if before_time else None
+
+        user = get_user(username, ctx, self.blossom_api)
+
+        rate_response = self.blossom_api.get(
+            "submission/rate/",
+            params={
+                "completed_by__isnull": False,
+                "completed_by": get_user_id(user),
+                "complete_time__gte": from_str,
+                "complete_time__lte": until_str,
+                "page_size": 365,
+                "time_frame": "day",
+            },
+        )
+        if rate_response.status_code != 200:
+            raise BlossomException(rate_response)
+
+        rate_data = rate_response.json()["results"]
+        rate_df = pd.DataFrame.from_records(rate_data, columns=["date", "count"])
+        # Convert date strings to datetime objects
+        rate_df["date"] = rate_df["date"].apply(lambda x: parser.parse(x))
+        rate_df = rate_df.set_index("date")
+
+        # Add the week number
+        rate_df["week"] = rate_df.index.to_series().apply(lambda x: _get_week_index(x))
+        # Add the week day
+        rate_df["day"] = rate_df.index.to_series().apply(lambda x: x.isocalendar()[2])
+
+        # All possible weeks, in case some are missing in the data
+        all_week_indexes = [
+            _get_week_index((before_time or start) - timedelta(weeks=weeks))
+            for weeks in range(53)

```

Comment:
Because technically there isn't exactly 53 weeks in a year, would rounding the weeks make an issue at all?

Reply:
For the purposes of general math, assuming 52 weeks is reasonable. In this case, it's not that we're rounding up, it's that we actually want to get the current week _and_ a year's worth of weeks before the current week. If we look at `_get_week_index()`, we see:

```python
def _get_week_index(date: datetime) -> int:
    """Get an identification number for the week.
    year * 100 + week
    """
    calendar = date.isocalendar()
    return calendar[0] * 100 + calendar[1]
```

We know from the type hints that we need to pass in a `datetime` object and we'll get an integer in return, so let's call it and see what happens.

```python
Python 3.10.2 (main, Mar  2 2022, 16:28:45) [Clang 13.0.0 (clang-1300.0.29.30)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> def _get_week_index(date):
...     calendar = date.isocalendar()
...     return calendar[0] * 100 + calendar[1]
...
>>> from datetime import datetime
>>> _get_week_index(datetime.now())
202233
```

Looks like the year and the index for the month. Now that we know that, we can plug in the rest, though it's worth noting that this is a compacted call.

```python
>>> [_get_week_index(datetime.now() - timedelta(weeks=count)) for count in range(53)]
[202233, 202232, 202231, 202230, 202229, 202228, 202227, 202226, 202225, 202224, 202223, 202222, 202221, 202220, 202219, 202218, 202217, 202216, 202215, 202214, 202213, 202212, 202211, 202210, 202209, 202208, 202207, 202206, 202205, 202204, 202203, 202202, 202201, 202152, 202151, 202150, 202149, 202148, 202147, 202146, 202145, 202144, 202143, 202142, 202141, 202140, 202139, 202138, 202137, 202136, 202135, 202134, 202133]
```

The above is functionally the same as:

```python
>>> week_indexes = []
>>> for count in range(53):
...   new_index = _get_week_index(datetime.now() - timedelta(weeks=count)
...   week_indexes.append(new_index)
...
>>> week_indexes
[202233, 202232, 202231, 202230, ...]
```

Checking the image in the original post, it looks like what we want to display is "the current week, no matter how many days are in it, and the entire previous year". Given that, this is the logic that powers how much time we actually populate the graph with.

Reply:
Yea, it's probably not completely accurate in all cases, but "good enough" :D